### PR TITLE
docs: sphinx: Document Java and C++ behaviour differences

### PR DESCRIPTION
**Pull request by rleigh-codelibre**
https://trello.com/c/s6HdQNWu/82-document-differences-between-c-and-java-pixel-data-handling

Testing: Check docs

### CODE REVIEW

**Comment by mtbc** [comment]
Where are the docs? http://www.openmicroscopy.org/site/support/contributing/ci-docs.html gets me to places like http://www.openmicroscopy.org/site/support/ome-model-staging/ but I remain a bit lost I'm afraid.

File: docs/sphinx/conversion.rst [hunk 1]
```diff
@@ -814,3 +814,66 @@ complex type, etc.), even when the logic would be identical for
...
+Pixel data buffering
+^^^^^^^^^^^^^^^^^^^^
+
+Pixel data is handled differently between the Java and C++
+implementations.  The primary reason for the difference is that the
+Java code uses raw ``byte[]`` arrays to contain pixel data.  This
+could not be implemented in C++ due to the limitations of C++ arrays
+discussed above, as well as having a number of additional limitations:
+
+- they aren't using the native pixel type, requiring conversion to the
```
Comment (mtbc) [inline docs/sphinx/conversion.rst]: "are not" (we tend to avoid contractions)

File: docs/sphinx/conversion.rst [hunk 2]
```diff
@@ -814,3 +814,66 @@ complex type, etc.), even when the logic would be identical for
...
+  required type, potentially also including endian conversion
+- they are unstructured, having no dimension ordering or dimension
+  size information
+
+The solution was to create a dedicated :cpp:class:`PixelBuffer`
```
Comment (mtbc) [inline docs/sphinx/conversion.rst]: "is"?

File: docs/sphinx/conversion.rst [hunk 3]
```diff
@@ -814,3 +814,66 @@ complex type, etc.), even when the logic would be identical for
...
+assumes it will never receive interleaved data will need to be updated
+to cope with it when porting to C++.
+
+The Java TIFF writer will always set interleaving if the number of
+samples per pixel is one (which is the recommended behaviour),
+overriding :cpp:func:`FormatWriter::setInterleaved()`.  The C++ TIFF
+writer will always set interleaving based upon
+:cpp:func:`FormatWriter::setInterleaved()`, and will not override the
+request of the caller.  This discrepancy will be rectified in a future
+release to match the behaviour of the Java reader; in practice there
```
Comment (mtbc) [inline docs/sphinx/conversion.rst]: "behavior" :smiley:

**Comment by mtbc** [comment]
Aha, thanks to @sbesson, @hflynn have now found this staged at https://www.openmicroscopy.org/site/support/ome-files-cpp-staging/ome-files/manual/html/conversion.html.

File: docs/sphinx/conversion.rst (hunk 2 above)
Comment (rleigh-codelibre) [inline docs/sphinx/conversion.rst]: This is describing what we did in the past several years back, so I'm not sure the present tense is appropriate.

Comment (mtbc) [inline docs/sphinx/conversion.rst]: Ah, it's no longer what happens? Fair enough, ta.

**Comment by rleigh-codelibre** [comment]
I've pushed fixes for the suggested corrections except for one which I've commented on.

File: docs/sphinx/conversion.rst (hunk 2 above)
Comment (rleigh-codelibre) [inline docs/sphinx/conversion.rst]: It's still used today, but the porting/conversion work is historical.

**Comment by sbesson** [comment]
@mtbc: are you fine with the current state of this PR?

**Comment by mtbc** [comment]
Will relist it for today if it's not already there.

File: docs/sphinx/conversion.rst [hunk 4]
```diff
@@ -814,3 +814,66 @@ complex type, etc.), even when the logic would be identical for
...
+assumptions may require some attention.
+
+The Java TIFF reader classes'
+:cpp:func:`FormatReader::isInterleaved()` method will always return
+``false``, irrespective of the TIFF ``PlanarConfiguration`` tag.  As a
+result, :cpp:func:`FormatReader::openBytes()` will always return pixel
+data with samples on separate contiguous planes.  In contrast, the C++
+TIFF reader classes' :cpp:func:`FormatReader::isInterleaved()` method
+will return ``true`` if the TIFF ``PlanarConfiguration`` is ``CONTIG``
+and ``false`` if ``SEPARATE``, and the
```
Comment (mtbc) [inline docs/sphinx/conversion.rst]: is this meant to be "SEPARATE is false"?

File: docs/sphinx/conversion.rst [hunk 5]
```diff
@@ -814,3 +814,66 @@ complex type, etc.), even when the logic would be identical for
...
+:cpp:func:`FormatReader::openBytes()` method will return pixel data
+with the appropriate interleaving, matching the same format in the
+TIFF file.  The Java behavior is due to the implementation details of
+its TIFF reading code; the C++ code uses libtiff and simply passes
+back the pixel data without any rearrangement.  Java code which
+assumes it will never receive interleaved data will need to be updated
+to cope with it when porting to C++.
```
Comment (mtbc) [inline docs/sphinx/conversion.rst]: out of curiosity, is it pretty trivial in C++ to de-interleave the data? if so, might be worth a mention

Comment (rleigh-codelibre) [inline docs/sphinx/conversion.rst]: Added at the end.

File: docs/sphinx/conversion.rst (hunk 4 above)
Comment (rleigh-codelibre) [inline docs/sphinx/conversion.rst]: false if PlanarConfiguration is SEPARATE is how it's intended to read

Comment (mtbc) [inline docs/sphinx/conversion.rst]: Ahh, ok.

File: docs/sphinx/conversion.rst [hunk 6]
```diff
@@ -814,3 +814,75 @@ complex type, etc.), even when the logic would be identical for
...
+  the required type, potentially also including endian conversion
+- they are unstructured, having no dimension ordering or dimension
+  size information
+
+The solution was to create a dedicated :cpp:class:`PixelBuffer`
+template class which could represent pixels of any type.  This is
+contained by a :cpp:class:`VariantPixelBuffer` class which can contain
+any of the supported pixel types.  This is therefore both flexible and
+strongly-typed.  The C++ code is slightly more complex as a result,
+but is safer, and the buffer can be passed around without the need for
```
Comment (simleo) [inline docs/sphinx/conversion.rst]: I'd replace "but is safer" with "but it is safer"

**Comment by dgault** [comment]
Not necessarily for this PR but it would probably be good at some point to have just a short bit of example code for the interleaving to point towards. The description is perfectly fine for now though.

Otherwise the text reads fine and makes sense to me.

**Comment by rleigh-codelibre** [comment]
We can add this to the tutorial samples in the following section.

**Comment by sbesson** [comment]
Thanks. The last changes make sense and the follow-up tutorial sample has been collected as https://trello.com/c/citrfg1i/2-doc-add-example-code-for-the-interleaving. Merging for 0.3.0

**Pull request merged**